docs/developer_overview.md

```diff
@@ -0,0 +1,527 @@
+
+`cub::ThreadReduce` 
+   - A normal function invoked and executed sequentially by a single thread that returns a valid result on that thread
+   - Single thread functions are usually an implementation detail and not exposed in CUB's public API
+
+`cub::WarpReduce` and `cub::BlockReduce`
+   - A "cooperative" function where threads concurrently invoke the same function to execute parallel work
+   - The function's return value is well-defined only on the "first" thread (lowest thread index)
+
+`cub::DeviceReduce` 
+   - A normal function invoked by a single thread that spawns additional threads to execute parallel work
+   - Result is stored in the pointer provided to the function
+   - Function returns a `cudaError_t` error code
+
+
+The table below provides a summary of these functions:
+
+| layer                | coop invocation | coop execution | max threads | valid result in |
+| -------------------- | --------------- | -------------- | ----------- | --------------- |
+| `cub::ThreadReduce`  | -               | -              | 1           | invoking thread |
+| `cub::WarpReduce`    | +               | +              | 32          | main thread     |
+| `cub::BlockReduce`   | +               | +              | 1024        | main thread     |
+| `cub::DeviceReduce ` | -               | +              | ∞           | global memory   |
+
+The details of how each of these layers are implemented is described below. 
+## Thread-scope
+
+Thread-scope functionality is usually presented with functions and rarely exposed to the user:
+
+```cpp
+template <int LENGTH, typename T, typename ReductionOp>
+__device__ __forceinline__ 
+T ThreadReduce(T* input,         
+               ReductionOp reduction_op,  
+               T prefix,     
+               Int2Type<LENGTH> /*length*/)
+{
+    T retval = prefix;
+
+    #pragma unroll
+    for (int i = 0; i < LENGTH; ++i)
+        retval = reduction_op(retval, input[i]);
+
+    return retval;
+}
+```
+
+## Warp-scope
+
+#### Overview
+
+Warp-scope functionality is provided by structures:
+
+```cpp
+template <typename T, 
+          int LOGICAL_WARP_THREADS = 32,
+          int LEGACY_PTX_ARCH = 0>
+class WarpReduce {
+  // ...
+  // (1)   define `_TempStorage` type
+  // ...
+  _TempStorage &temp_storage;
+public:
+    
+  // (2)   wrap `_TempStorage` in uninitialized memory
+  struct TempStorage : Uninitialized<_TempStorage> {};
+
+  __device__ __forceinline__ WarpReduce(TempStorage &temp_storage)
+  /* (3)   reinterpret cast */
+    : temp_storage(temp_storage.Alias()) 
+  {}
+
+  /* (4)   actual algorithms */
+  __device__ __forceinline__ T Sum(T input);
+};
+```
+
+To use warp-scope algorithms, warp size has to be known at compilation time (provided as a template parameter). It's possible to diverge from using the default warp size, in this case:
```

Comment:
Can we add a sentence that clearly advices against non power of two threads

Reply:
Why advice against it if there's a use case?